 qpxtool (0.7.2-4.1-ArtKarMOD) unstable; urgency=medium
 .
   * support for BD-RE WH16NS58
   * other changes
Author: ArtKar <artkar@intercore.com.pl>

---
The information above should follow the Patch Tagging Guidelines, please
checkout http://dep.debian.net/deps/dep3/ to learn about the format. Here
are templates for supplementary fields that you might want to add:

Origin: <vendor|upstream|other>, <url of original patch>
Bug: <url in upstream bugtracker>
Bug-Debian: https://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: <no|not-needed|url proving that it has been forwarded>
Reviewed-By: <name and email of someone who approved the patch>
Last-Update: 2019-04-15

--- qpxtool-0.7.2-4.1.orig/README
+++ qpxtool-0.7.2-4.1/README
@@ -1,3 +1,17 @@
+QPxTool 0.7.2 modded by ArtKar v3
+ - toggle P95 graph / simple graph
+
+QPxTool 0.7.2 modded by ArtKar v2
+ - BDXL in SQL
+ - added support for DVD Quality test on BD-RE WH16NS58
+
+QPxTool 0.7.2 modded by ArtKar
+ - added support for BD-RE WH16NS58 ( only tested Blu Ray quality test )
+ - added support for multilayer Blu Ray discs
+ - added support for BDXL
+ - changed graph style
+
+
 ************************************************************

 QPxTool 0.7.2
--- qpxtool-0.7.2-4.1.orig/console/qscan/qscan.cpp
+++ qpxtool-0.7.2-4.1/console/qscan/qscan.cpp
@@ -628,7 +628,9 @@ int main(int argc, char** argv) {
 			}
 		}
 		printf(IMEDIA "Layers        : %d\n", dev->media.layers);
-
+		if ( dev->media.type & (DISC_BD) ) {
+			printf(IMEDIA "GB per Layer  : %d\n", dev->media.gbpl);
+		}
 		if ( dev->media.type & (DISC_DVD) ) {
 //			read_disc_regions(drive);
 			if (!dev->media.dvdcss.protection) {

--- qpxtool-0.7.2-4.1.orig/gui/include/device.h
+++ qpxtool-0.7.2-4.1/gui/include/device.h
@@ -344,6 +344,8 @@ struct MediaInfo {
 	QString		mid;
 	QString		layers;
 	int			ilayers;
+	QString			gbpl;
+	int			igbpl;
 	QString		prot;
 	QString		regions;

--- qpxtool-0.7.2-4.1.orig/gui/include/mainwidget.h
+++ qpxtool-0.7.2-4.1/gui/include/mainwidget.h
@@ -50,6 +50,7 @@ public slots:
 	void selectDevice();
 	void reconfig();
 	void setSidebarVisible(bool);
+	void setSimpleGraph(bool);
 	void selectTab(int);
 	//inline void reconfig() { emit configured(); };

--- qpxtool-0.7.2-4.1.orig/gui/include/mainwindow.h
+++ qpxtool-0.7.2-4.1/gui/include/mainwindow.h
@@ -167,6 +167,7 @@ private:
 				*act_about;

 	QAction		    *act_sb;
+	QAction		    *act_sg;
 	QList<QAction*> act_sblist;
 	QActionGroup    *act_sbgrp;

--- qpxtool-0.7.2-4.1.orig/gui/include/qpxsettings.h
+++ qpxtool-0.7.2-4.1/gui/include/qpxsettings.h
@@ -166,6 +166,7 @@ public:

 // general options
 	bool		show_sidebar;
+	bool		show_simplegraph;
 	bool		show_allctl;
 	bool		report_autosave;
 	QString		report_path;

--- qpxtool-0.7.2-4.1.orig/gui/src/db_report_selection.cpp
+++ qpxtool-0.7.2-4.1/gui/src/db_report_selection.cpp
@@ -228,7 +228,7 @@ void DbReportSelection::search()
 				reports \
 			WHERE \
 				label ILIKE '%"+elabel->text()+"%' \
-			ORDER BY datetime");
+			ORDER BY label");
 	if (!q->exec()) {
 		qDebug() << q->lastError().text();
 		goto err;
--- qpxtool-0.7.2-4.1.orig/gui/src/device.cpp
+++ qpxtool-0.7.2-4.1/gui/src/device.cpp
@@ -274,6 +274,8 @@ void device::clearMinfo()
 	media.erasable	= "-";
 	media.layers	= "-";
 	media.ilayers	= 1;
+	media.gbpl	= "-";
+	media.igbpl	= 25;
 	media.prot		= "-";
 	media.regions	= "-";
 	media.creads	= 0;
@@ -1335,6 +1337,7 @@ void device::qscan_process_line(QString&
 #ifndef QT_NO_DEBUG
 //			qDebug("|" + sl[0] + "|" + sl[1] + "|");
 #endif
+
 			if (sl[0].contains("Media type", Qt::CaseInsensitive)) {
 				if (sl[1].contains("No Media", Qt::CaseInsensitive)) {
 					media.type = "-";
@@ -1357,7 +1360,12 @@ void device::qscan_process_line(QString&
 				media.layers = sl[1];
 				media.ilayers = sl[1].toInt();
 				if (media.ilayers <= 0)
-					media.ilayers = 1;
+					media.ilayers = 1;
+			} else if (sl[0].contains("GB per Layer", Qt::CaseInsensitive)) {
+				media.gbpl = sl[1];
+				media.igbpl = sl[1].toInt();
+				if (media.igbpl <= 0)
+					media.igbpl = 25;
 			} else if (sl[0].contains("Protection", Qt::CaseInsensitive)) {
 				media.prot = sl[1];
 			} else if (sl[0].contains("Regions", Qt::CaseInsensitive)) {
--- qpxtool-0.7.2-4.1.orig/gui/src/mainwidget.cpp
+++ qpxtool-0.7.2-4.1/gui/src/mainwidget.cpp
@@ -148,6 +148,7 @@ QPxMainWidget::~QPxMainWidget()
 }

 void QPxMainWidget::setSidebarVisible(bool en) { settings->show_sidebar = en; bframe->setVisible(en); }
+void QPxMainWidget::setSimpleGraph(bool en) { settings->show_simplegraph = en; }
 void QPxMainWidget::selectTab(int idx) { grp->button(idx)->setChecked(true); stack->setCurrentIndex(idx); }
 void QPxMainWidget::reconfig() { emit configured(); }
 void QPxMainWidget::clearDev() { tab_MediaInfo->clear(); tab_DevInfo->clear(); }
--- qpxtool-0.7.2-4.1.orig/gui/src/mainwindow.cpp
+++ qpxtool-0.7.2-4.1/gui/src/mainwindow.cpp
@@ -221,6 +221,7 @@ void QPxToolMW::winit()
 	layout->addWidget(mwidget);

 	connect(act_sb, SIGNAL(toggled(bool)), mwidget, SLOT(setSidebarVisible(bool)));
+	connect(act_sg, SIGNAL(toggled(bool)), mwidget, SLOT(setSimpleGraph(bool)));

 	connect(pb_loej, SIGNAL(clicked()), this, SLOT(loejToggle()));
 	connect(pb_lock, SIGNAL(clicked()), this, SLOT(lockToggle()));
@@ -259,6 +260,11 @@ void QPxToolMW::create_actions()
 	act_sb->setCheckable(true);
 	act_sb->setChecked(set.show_sidebar);

+	act_sg     = new QAction(QIcon(":images/"), tr("Show simple graph"), this);
+	act_sg->setShortcut( QKeySequence("Alt+G") );
+	act_sg->setCheckable(true);
+	act_sg->setChecked(set.show_simplegraph);
+
 	QAction *act;
 	act_sbgrp = new QActionGroup(this);

@@ -341,6 +347,8 @@ void QPxToolMW::winit_menubar()
 	menu->addActions(act_sblist);
 	menu->addSeparator();
 	menu->addAction(act_sb);
+	menu->addSeparator();
+	menu->addAction(act_sg);

 	menubar->addMenu(menu);

--- qpxtool-0.7.2-4.1.orig/gui/src/qpxgraph.cpp
+++ qpxtool-0.7.2-4.1/gui/src/qpxgraph.cpp
@@ -30,8 +30,8 @@
 #define MARGIN_DEFR	40
 #define MARGIN_DEFB	18

-#define GRID_STYLE Qt::DotLine
-//#define GRID_STYLE Qt::DashLine
+//#define GRID_STYLE Qt::DotLine
+#define GRID_STYLE Qt::DashLine

 class IntList : public QList<int>
 {
@@ -307,7 +307,7 @@ void QPxGraph::drawGraph(QPainter *p, QS
 			HscaleLBA = (1<<19) * 5 * dev->media.ilayers/sg.width();
 			Vscale1X = Vscale * 3;
 		} else if (dev->media.type.startsWith("BD")) {
-			HscaleLBA = (1<<19) * 25 * dev->media.ilayers/sg.width();
+			HscaleLBA = (1<<19) * dev->media.igbpl * dev->media.ilayers/sg.width();
 			Vscale1X = Vscale * 4;
 		}

@@ -504,7 +504,9 @@ void QPxGraph::drawErrc(QPainter* p, con
 				x = (int) (dev->testData.errc[i].raw.lba/HscaleLBA);
 				if (dev->testData.errc[i].raw.err[e] >=0 ) {
 #ifdef SHOW_P95ERRC
+if (!(settings->show_simplegraph)) {
 					xerr.append(dev->testData.errc[i].raw.err[e]);
+}
 #endif
 // update min/max
 					if (min < 0 || min > dev->testData.errc[i].raw.err[e])
@@ -516,17 +518,23 @@ void QPxGraph::drawErrc(QPainter* p, con
 				if (x!=xo || i==(dev->testData.errc.size()-1)) {
 // min-max
 					p->setPen(QPen(*settings->col_errc.raw[e], 1));
+if (settings->show_simplegraph) {
+                                        p->drawLine(xo, errc2h(s.height(), 0),
+                                                                xo, errc2h(s.height(), max));
+} else {
 					p->drawLine(xo, errc2h(s.height(), min),
 								xo, errc2h(s.height(), max));
-
+}
 // P=0.95
 #ifdef SHOW_P95ERRC
+if (!(settings->show_simplegraph)) {
 					M = (int)xerr.M();
 					D = (int)sqrt(xerr.dispers(M));
 					p->setPen(QPen(settings->col_errc.raw[e]->darker(), 1));
 					p->drawLine(xo, errc2h(s.height(), M-D),
 								xo, errc2h(s.height(), M+D));
 					xerr.clear();
+}
 #endif
 					min = -1;
 					max = -1;
@@ -866,7 +874,7 @@ void QPxGraph::drawGrid(QPainter* p, con
 		GBperLayer = 5;
 	} else if (dev->media.type.startsWith("BD")) {
 		isCD = 0;
-		GBperLayer = 25;
+		GBperLayer = dev->media.igbpl;
 	}

 	if (isCD) {
--- qpxtool-0.7.2-4.1.orig/gui/src/qpxsettings.cpp
+++ qpxtool-0.7.2-4.1/gui/src/qpxsettings.cpp
@@ -37,6 +37,7 @@ QPxSettings::QPxSettings()
 //	geometry_testsel = QRect(0,0,0,0);

 	show_sidebar  = 0;
+	show_simplegraph  = 0;
 	show_allctl   = 0;
 	actions_flags = 0;

@@ -76,6 +77,7 @@ QPxSettings& QPxSettings::operator = (co
 	geometry_pref = o.geometry_pref;

 	show_sidebar    = o.show_sidebar;
+	show_simplegraph    = o.show_simplegraph;
 	show_allctl     = o.show_allctl;
 	report_autosave = o.report_autosave;
 	report_path     = o.report_path;
@@ -135,6 +137,7 @@ void QPxSettings::load()
 	settings = new QSettings(QSettings::NativeFormat, QSettings::UserScope, "QPxTool", "qpxtool");
 	settings->beginGroup("/common");
 		show_sidebar     = settings->value("show_sidebar", 0).toBool();
+		show_simplegraph     = settings->value("show_simplegraph", 0).toBool();
 		show_allctl      = settings->value("show_allctl", 0).toBool();
 		report_autosave  = settings->value("report_autosave", 0).toBool();
 		report_path      = settings->value("report_path", "").toString();
@@ -204,6 +207,7 @@ void QPxSettings::save()
 	settings = new QSettings(QSettings::NativeFormat, QSettings::UserScope, "QPxTool", "qpxtool");
 	settings->beginGroup("/common");
 		settings->setValue("show_sidebar", show_sidebar);
+		settings->setValue("show_simplegraph", show_simplegraph);
 		settings->setValue("show_allctl", show_allctl);
 		settings->setValue("report_autosave", report_autosave);
 		settings->setValue("report_path", report_path);
--- qpxtool-0.7.2-4.1.orig/gui/src/resultsio.cpp
+++ qpxtool-0.7.2-4.1/gui/src/resultsio.cpp
@@ -203,7 +203,9 @@ void ResultsReader::run() {
 							dev->media.grec    = attr.value("gigarec").toString().toDouble();
 							dev->media.spd1X   = attr.value("spd1X").toString().toInt();
 							dev->media.layers  = attr.value("layers").toString();
+							dev->media.gbpl  = attr.value("gbpl").toString();
 							dev->media.erasable= attr.value("erasable").toString();
+							dev->media.igbpl = dev->media.gbpl.toInt();
 							dev->media.ilayers = dev->media.layers.toInt();
 							dev->media.dstate  = attr.value("dstate").toString();
 							dev->media.sstate  = attr.value("sstate").toString();
@@ -396,6 +398,7 @@ void ResultsWriter::run() {

 	xml.writeAttribute("spd1X",     QString::number(dev->media.spd1X));
 	xml.writeAttribute("layers",    QString::number(dev->media.ilayers));
+	xml.writeAttribute("gbpl",    QString::number(dev->media.igbpl));
 	xml.writeAttribute("erasable",  dev->media.erasable);

 	xml.writeAttribute("dstate", dev->media.dstate);
--- qpxtool-0.7.2-4.1.orig/lib/include/qpx_mmc.h
+++ qpxtool-0.7.2-4.1/lib/include/qpx_mmc.h
@@ -144,6 +144,7 @@ typedef	struct {
 	uint8_t		disc_size;	// indicates 120/80mm disc
 	uint8_t		polarity;   // Push-Pull polarity flags per layer for BD (indicates HtL or LtH)
 	uint8_t		layers;		// Layers num (!CD)
+	uint8_t		gbpl;		// Layers num (!CD)
 	int			sectsize;
 	int32_t	capacity;	// Recorded capacity in sectors
 	msf		capacity_msf;
--- qpxtool-0.7.2-4.1.orig/lib/qpxtransport/qpx_mmc.cpp
+++ qpxtool-0.7.2-4.1/lib/qpxtransport/qpx_mmc.cpp
@@ -2316,13 +2316,21 @@ int determine_disc_type(drive_info* driv
 		} else if (drive->media.type & DISC_BD) {
  			drive->rd_buf[4]=0;
 			drive->cmd[0] = MMC_READ_DVD_STRUCTURE;
-			drive->cmd[7] = 0;//0x11; //dvd_dash;
-			drive->cmd[9] = 36;
-			drive->cmd[11] = 0;
+			drive->cmd[1] = 1;
+			drive->cmd[7] = 0;
+			drive->cmd[9] = 66;
+
+
 			if ((drive->err=drive->cmd.transport(READ,drive->rd_buf,36)))
 				if (!drive->silent) sperror ("READ_DVD_STRUCTURE",drive->err);
 			drive->media.book_type = 0;
-			drive->media.layers = 1 + ((drive->rd_buf[6] & 0x60) >> 5);
+			drive->media.layers = 0 + ((drive->rd_buf[16] & 0xF0) >> 4);
+			if ((drive->rd_buf[17] & 0x0F) == 0) drive->media.gbpl = 25;
+			else if ((drive->rd_buf[17] & 0x0F) == 1) drive->media.gbpl = 25;
+			else if ((drive->rd_buf[17] & 0x0F) == 2) drive->media.gbpl = 27;
+			else if ((drive->rd_buf[17] & 0x0F) == 5) drive->media.gbpl = 33;
+
+
 			read_mediaid_bd(drive);
 			if (!drive->silent) printf("** MID: '%s'\n",drive->media.MID);
 		}
@@ -2691,7 +2699,10 @@ int set_cd_speed(drive_info* drive) {
 	drive->cmd[11] = 0;
 	if ((drive->err=drive->cmd.transport(NONE,NULL,0) )) {
 //		if (drive->err != 0x23A02) drive->capabilities&=(NCAP_SET_CD_SPEED);
-		if (!drive->silent) sperror ("SET_CD_SPEED",drive->err); return (drive->err);
+		if (!drive->silent) {
+			sperror ("SET_CD_SPEED",drive->err);
+			return (drive->err);
+		}
 	}
 	return 0;
 }
@@ -3119,7 +3130,10 @@ int plextor_px755_get_auth_code(drive_in
 		{ if (!dev->silent) sperror ("PLEXTOR_PX755_GET_AUTH_CODE",dev->err); return dev->err;}
 	if (!dev->silent) {
 		printf("** Get PX755 auth: ");
-		for (int i=0; i<16; i++) printf("0x%02X ",dev->rd_buf[i]&0xFF); printf("\n");
+		for (int i=0; i<16; i++) {
+			printf("0x%02X ",dev->rd_buf[i]&0xFF);
+			printf("\n");
+		}
 	}
 	return 0;
 }
--- qpxtool-0.7.2-4.1.orig/plugins/liteon/qscan_cmd_liteon.cpp
+++ qpxtool-0.7.2-4.1/plugins/liteon/qscan_cmd_liteon.cpp
@@ -305,26 +305,32 @@ int scan_liteon::cmd_cd_errc_block(cd_er
 // ********************** DVD ERRC commands
 int scan_liteon::cmd_dvd_errc_block(dvd_errc *data)
 {
+	bool retry = false;
+	if (!lba) {
+		retry = true;
+		// if first sector scan requested
+		// we have to seek to first sector
+		dev->cmd[0] = MMC_SEEK;
+		if ((dev->err=dev->cmd.transport(READ,dev->rd_buf,2048))){
+			sperror ("READ",dev->err); return 1;
+		}
+	}
+
+dvd_errc_repeat:
 	dev->cmd[0] = 0xF3;
 	dev->cmd[1] = 0x0E;
 	dev->cmd[8] = 0x10;
 	dev->cmd[11]= 0x00;
-	if ((dev->err=dev->cmd.transport(READ,dev->rd_buf,10))){
-		sperror ("LiteOn_errc_dvd_read_block",dev->err); return 1;
+	if ((dev->err=dev->cmd.transport(READ,dev->rd_buf,0x10))) {
+		sperror ("LiteOn_errc_bd_read_block",dev->err); return 1;
 	}
-#if 0
-	for (int i=0; i<10; i++) {
-	    printf(" %02X",dev->rd_buf[i]);
-	}
-	printf("\n");
-#endif
-
-// Data Received:
-// 00000000  00 00 00 8E 00 00 00 00                           ...Ž....
-
-//	lba = ((dev->rd_buf[1] << 16 )& 0xFF0000) + ((dev->rd_buf[2] << 8)&0xFF00 ) + (dev->rd_buf[3] & 0xFF);
 	lba = ntoh32(dev->rd_buf);

+	if (!lba && retry) {
+		retry = false;
+		goto dvd_errc_repeat;
+	}
+
 	data->pie = ntoh16(dev->rd_buf+4);
 	data->pif = ntoh16(dev->rd_buf+6);
 	data->poe = 0;
--- qpxtool-0.7.2-4.1.orig/plugins/liteon/qscan_plugin_liteon.h
+++ qpxtool-0.7.2-4.1/plugins/liteon/qscan_plugin_liteon.h
@@ -112,14 +112,13 @@ static const drivedesclist drivelist =

 	{ "SONY    ",  DEV_LITEON, "DVD RW DW-Q58A",	LTN_SDVDR_G3, CHK_ERRC_CD | CHK_ERRC_DVD },
 	{ "SONY    ",  DEV_LITEON, "DVD RW DW-Q60A",	LTN_SDVDR_G3, CHK_ERRC_CD | CHK_ERRC_DVD },
-
+        { "HL-DT-ST",  DEV_LITEON, "BD-RE  WH16NS58 ",	LTN_BDR, CHK_ERRC_CD | CHK_ERRC_DVD | CHK_ERRC_BD },
 	{ "", 0, "", 0}
 };

 static const drivedesclist blacklist =
 {
 	{ "TEAC    ",  DEV_TEAC, "CD-W552E", 0 },
-	{ "HL-DT-ST",  DEV_LG,   "",		 0 },
 	{ "TSSTcorp",  DEV_TSST, "",		 0 },

 	{ "", 0, "", 0}
